playlistmanager.lua

```diff
@@ -1071,63 +1074,113 @@ end
-            if res.status == 0 then
-              local json, err = utils.parse_json(res.stdout)
-              if not err then
-                local is_playlist = json['_type'] and json['_type'] == 'playlist'
-                local title = (is_playlist and '[playlist]: ' or '') .. json['title']
-                msg.verbose(filename .. " resolved to '" .. title .. "'")
-                url_table[filename] = title
-                refresh_globals()
-                if playlist_visible then showplaylist() end
-                return
-              else
-                msg.error("Failed parsing json, reason: "..(err or "unknown"))
-              end
             else
-              msg.error("Failed to resolve url title "..filename.." Error: "..(res.error or "unknown"))
+              msg.error("Failed parsing json, reason: "..(err or "unknown"))
             end
-          end)
+          else
+            msg.error("Failed to resolve url title "..filename.." Error: "..(res.error or "unknown"))
+          end
+         end)
+
+        mp.add_timeout(settings.resolve_title_timeout, function()
+        mp.abort_async_command(req)
+      end)
+    end
+
+    local function resolve_local()
+    local args = { "ffprobe", "-show_format", "-show_entries", "format=tags", "-loglevel", "quiet", filename }
+    local req = mp.command_native_async(
+      {
+        name = "subprocess",
+        args = args,
+        playback_only = false,
+        capture_stdout = true
+      }, function (success, res)
+          if res.killed_by_us then
+            msg.verbose('Request to resolve title ' .. filename .. ' timed out')
+            return
+          end
+
+          local title = string.match(res.stdout, "title=([^\n\r]+)")
+
+          if not title then
+      	    return
+          else
+            msg.verbose(filename .. " resolved to '" .. title .. "'")
+            title_table[filename] = title
+            refresh_globals()
+            if playlist_visible then showplaylist() end
+              return
+          end
+        end)
 
-      mp.add_timeout(settings.resolve_title_timeout, function()
+        mp.add_timeout(settings.resolve_title_timeout, function()
         mp.abort_async_command(req)
       end)
+    end
 
+    if i ~= pos
+      and settings.resolve_titles
+      and settings.prefer_titles == 'url'
+      and filename
+      and filename:match('^https?://')
+      and not title
+      and not title_table[filename]
+      and not requested_titles[filename]
+    then
+      requested_titles[filename] = true
+      resolve_url()
+    elseif i ~= pos
+      and settings.prefer_titles == 'all'
+      and filename
+      and not title
+      and not title_table[filename]
+      and not requested_titles[filename]
+    then
+      requested_titles[filename] = true
+      if settings.resolve_titles and filename:match('^https?://') then
+        resolve_url()
+      elseif settings.resolve_local_titles then
+        resolve_local()
+      end
```

Comment:
There is probably some easier way to do this. Perhaps an early return for non matching cases, then following checks could be more simple.